src/Linux/Native/RunAsUser/main.cpp

```diff
@@ -0,0 +1,209 @@
+/* ****************************************************************************
+*
+* Copyright (c) Microsoft Corporation. All rights reserved.
+*
+*
+* This file is part of Microsoft R Host.
+*
+* Microsoft R Host is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Microsoft R Host is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Microsoft R Host.  If not, see <http://www.gnu.org/licenses/>.
+*
+* ***************************************************************************/
+
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <pwd.h>
+#include <unistd.h>
+#include <string>
+#include <mutex>
+#include <security/pam_appl.h>
+#include <security/pam_misc.h>
+#include <boost/endian/buffers.hpp>
+
+#include "picojson.h"
+#include "util.h"
+
+FILE *input = nullptr, *output = nullptr, *error = nullptr;
+
+#define RTVS_AUTH_OK          0
+#define RTVS_AUTH_INIT_FAILED 200
+#define RTVS_AUTH_BAD_INPUT   201
+#define RTVS_AUTH_NO_INPUT    202
+
+std::string read_string(FILE* stream) {
+    boost::endian::little_uint32_buf_t data_size;
+    if (fread(&data_size, sizeof data_size, 1, stream) != 1) {
+        return std::string();
+    }
+
+    std::string str(data_size.value(), '\0');
+    if (!str.empty()) {
+        if (fread(&str[0], str.size(), 1, stream) != 1) {
+            return std::string();
+        }
+    }
+
+    return str;
+}
+
+std::mutex output_lock;
+void write_string(FILE* stream, std::string &data) {
+    boost::endian::little_uint32_buf_t data_size(static_cast<uint32_t>(data.size()));
+
+    std::lock_guard<std::mutex> lock(output_lock);
+    if (fwrite(&data_size, sizeof data_size, 1, stream) == 1) {
```

Comment:
There should probably be some error handling here. Since we can't really do much if the writes failed, I think you can just call std::terminate to fail fast.